src/custom_slack_provider/helpers.py

```diff
@@ -0,0 +1,202 @@
+            # is not supported. Issue is that one cannot simply
+            # remove the social account from the other user, as
+            # that may render the account unusable.
+            level = messages.ERROR
+            message = 'socialaccount/messages/account_connected_other.txt'
+        else:
+            # This account is already connected -- we give the opportunity
+            # for customized behaviour through use of a signal.
+            action = 'updated'
+            message = 'socialaccount/messages/account_connected_updated.txt'
+            signals.social_account_updated.send(
+                sender=SocialLogin,
+                request=request,
+                sociallogin=sociallogin)
+    else:
+        # New account, let's connect
+        action = 'added'
+        sociallogin.connect(request, request.user)
+        signals.social_account_added.send(sender=SocialLogin,
+                                          request=request,
+                                          sociallogin=sociallogin)
+    default_next = get_adapter(request).get_connect_redirect_url(
+        request,
+        sociallogin.account)
+    next_url = sociallogin.get_redirect_url(request) or default_next
+    get_account_adapter(request).add_message(
+        request, level, message,
+        message_context={
+            'sociallogin': sociallogin,
+            'action': action
+        }
+    )
+    return HttpResponseRedirect(next_url)
+
+
+def complete_social_login(request, sociallogin):
+    assert not sociallogin.is_existing
+    sociallogin.lookup()
+    try:
+        get_adapter(request).pre_social_login(request, sociallogin)
+        signals.pre_social_login.send(sender=SocialLogin,
+                                      request=request,
+                                      sociallogin=sociallogin)
+        process = sociallogin.state.get('process')
+        if process == AuthProcess.REDIRECT:
+            return _social_login_redirect(request, sociallogin)
+        elif process == AuthProcess.CONNECT:
+            return _add_social_account(request, sociallogin)
+        else:
+            return _complete_social_login(request, sociallogin)
+    except ImmediateHttpResponse as e:
+        return e.response
+
+
+def _social_login_redirect(request, sociallogin):
+    next_url = sociallogin.get_redirect_url(request) or '/'
+    return HttpResponseRedirect(next_url)
+
+
+def _complete_social_login(request, sociallogin):
+    if request.user.is_authenticated:
+        get_account_adapter(request).logout(request)
+    if sociallogin.is_existing:
+        # Login existing user
+        ret = _login_social_account(request, sociallogin)
+        signals.social_account_updated.send(
+            sender=SocialLogin,
+            request=request,
+            sociallogin=sociallogin)
+    else:
+        # New social user
+        ret = _process_signup(request, sociallogin)
+        extra_data = sociallogin.account.extra_data
```

Comment:
Since most of the rest is just the same code I would add a comment here to say that this is custom code and what is happening below.

Reply:
Done